pyi.py

```diff
@@ -917,8 +920,39 @@ def _check_union_members(self, members: Sequence[ast.expr]) -> None:
                 dupes_in_union = True
 
         if not dupes_in_union:
+            self._check_for_redundant_numeric_unions(members)
             self._check_for_multiple_literals(members)
 
+    def _Y041_error(
+        self, members: Sequence[ast.expr], subtype: str, supertype: str
+    ) -> None:
+        self.error(
+            members[0],
+            Y041.format(implicit_subtype=subtype, implicit_supertype=supertype),
+        )
+
+    def _check_for_redundant_numeric_unions(self, members: Sequence[ast.expr]) -> None:
+        complex_in_union, float_in_union, int_in_union = False, False, False
+
+        for member in members:
+            if _is_complex(member):
+                complex_in_union = True
+            elif _is_float(member):
+                float_in_union = True
+            elif _is_int(member):
+                int_in_union = True
+
+            if all((complex_in_union, float_in_union, int_in_union)):
+                break
```

Comment:
Why the early break out here?

Reply:
The algorithm is the same as `any()`, except that it checks for three conditions in the same loop.

`any()` breaks from the loop as soon as it establishes that there is at least one value in the iterable for which the condition is truthy.

In the same way, this algorithm breaks from the loop as soon as it has established that, for each of the three conditions, there is at least one element in the iterable for which the condition is truthy.

Reply:
I recommend to remove that early check here:

* It makes the code harder to understand.
* It's another code path that could introduce bugs if the code is ever changed.
* I don't think it will measurably improve performance. It could even degrade performance, because it introduces extra function calls in the common case, where no numeric types are part of the union. But we would need some metrics to prove either case.

Reply:
That all makes sense, I'll remove the check!